### DISCUSSION TITLE
Improve IPC interfaces

### DISCUSSION THREAD

**Discussion by gterzian** [servo/servo#23913] [General]
Currently, our IPC is pretty good for cases where two processes setup communication lines using a channel, and then use that channel for each operation at the level of their "main event-loop". 

An example is how script messages to the constellation. Each script-thread(process?) has a dedicated sender to the constellation, and vice-versa, and that unique channel is used for many operations. Those are all "top-level" operations, such as navigating and so on. 

Our IPC is also very convenient for "one-off" type of operations between processes, where one is able to register a kind of "callback" using the `ROUTER`, and not having to setup something using the main event-loop where that code happens to be running. 

An example is how script communicates with fetch. Each fetch is done by sending a message to the resource manager, however that message itself contains a sender to communicate results directly into some sort of "callback" that would have been setup on the router thread in the script-process. The callback always consist of then "translating" the fetch result into the appropriate task, and enqueuing it onto the event-loop thread. 

it's convenient because the callback is just a closure registered right from the DOM code, without  the need to do some orchestration with the script-thread or something "at the top" of the event-loop where the DOM code is running. 

However the problem is each callback requires a new channel, as well as sending the sender half across process, which results in overuse of underlying system resources. 

See 

https://github.com/servo/servo/issues/23905

and https://github.com/servo/ipc-channel/issues/240

With some sort of ongoing prototype at https://github.com/servo/servo/pull/23909

And a discussion that started below cc @asajeffrey

> sending the same ipc channel over ipc many times uses up an fd each time it's sent, so recycling the ipc channel doesn't help

I can also imagine the `IpcHandle` only including it's callback id, not an actual sender. 

Then each process using a handle, like fetch, could use a single sender to "message to the callback corresponding to the handle". 

So the current `SharedRouter` could be matched with a `SharedDispatcher` on the sending side, meaning each ipc callback would just be an id sent across process, with a callback stored on the "receiving" side. 

Then when the time comes for fetch to "call the callback", it would use the `SharedDispatcher`, combined with the `CallbackId` made available in a dedicated handle. 

The problem we're faced here might be less one of optimizing the underlying OS operations, and rather rationalizing the higher-level interfaces, so that we don't create a channel for each operation, and also avoid sending a clone of a sender across process for each.

_Originally posted by @gterzian in https://github.com/servo/servo/issues/23905#issuecomment-518071970_

**Comment by gterzian** [comment]
How would we setup the intial communication lines between such `SharedRouter` and `SharedDispatcher`? It could be done with the constellation, each time a new pipeline is created. 

Actually, we could perhaps even by pass the constellation, since script could just register its router with the resource manager by sending a message directly: `CoreResourceMsg::NewRouter(PipelineId, IpcSender<IpcCallbackMsg>)`. 

Then a `IpcHandle` would need to contain the following:

1. a `PipelineId`, so a `SharedDispatcher`, for example in fetch, would know which pipeline the handle related to and could send the message on the "right" sender, thereby avoiding the need for the handle itself to contain a sender.
2. a `CallbackId`, so the `SharedRouter`, upon receiving a message from the dispatcher, would know which callback to invoke with the data in the message. 

At that point, a given process could register many callbacks, only using a single ipc-channel, with the receiver used by the `SharedRouter`, and the sender shared initially to the `SharedDispatcher` in another process.

**Comment by gterzian** [comment]
As an update, it turned out to be possible to avoid sending a (clone-of) an ipc-sender across process for each callback, and that without introducing that many new concepts in the code(I think I went a bit overboard with the `SharedDispatcher` and so on). See https://github.com/servo/servo/pull/23909/commits/62940291ad69f3fdc18d2dcb1841f7b3211c6fc6

So the ipc-sender corresponding to the shrared-router of a pipeline is sent only once to the resource process, then when the process receives a new fetch command containing a `IpcHandle`, it sets a clone of the sender on the handle. This clone is then not ipc'ed anywhere, i't just used for the duration of a fetch and doesn't cross a process boundary. 

So the benefit is that the `IpcHandle`, which is created once per callback, and represents a way to "send a message to the callback",  doesn't contain an ipc-sender when sent across process to fetch. The sender is only set at the point of use inside the fetch process.

**Comment by asajeffrey** [comment]
I suspect we'd get the same benefit if we added a per-process lookup table to ipc channel which cached fds. That way we wouldn't use up an fd every time we recv'd a channel.

It's even worse than using an fd every time a channel crosses a process boundary, it uses an fd every time it is recv'd, even if it didn't cross a process boundary.

The fix for this is fairly straightforward (well, in that caching ever is) but it needs to be done in the ipc-channel crate rather than servo.

**Comment by gterzian** [comment]
> I suspect we'd get the same benefit if we added a per-process lookup table to ipc channel which cached fds. That way we wouldn't use up an fd every time we recv'd a channel.

I see, so that could be a way to fix the "we send a clone of the same sender repeatedly on a channel, and that currently creates an fd on each ipc recv(whether it crosses process or not)". 

It would still leave the problem of having to create an ipc-channel each time the `ROUTER` is used(for each websocket, each fetch, each image, each messageport, each stylesheet, and so on).  

Note that all crashes reported at https://github.com/servo/servo/issues/23905 where directly due to a call to `ipc::channel()`, I haven't seen one yet on a `recv`. Although receiving senders in a `recv` probably contributed to the overall fd count, it's interesting to see that it's creating a channel as part of setting up a route(or each time a fetch canceller is initialized) that is causing the crashes.

**Comment by gterzian** [comment]
I also wonder if would be optimal to make ipc-channel itself more complicated(who wants to handle the fd caching bugs that will come-in on the issue tracker?), as opposed to evolve what appears like an immature design on the Servo side of things?

If we add some glue code here, bugs will not affect users of ipc-channel generally. 

It might also be easier to debug, and further incrementally evolve, some glue code here, rather than an fd caching layer inside ipc-channel?

**Comment by gterzian** [comment]
Update for https://github.com/servo/servo/pull/23909

I have now also tried applying the `SharedRouter` mechanism to the font-cache, and it seems to work well. I've relaced the `PipelineId` with a dedicated `IpcRouterId`, since a router can now be tied to a component that isn't a pipeline. 

This also highlights how we need more threads to have a `PipelineNamespace` installed, which already came up for workers and message-ports at https://github.com/servo/servo/pull/23909

Essentially, the ability to create unique Ids based on a pipeline namespace seems really useful, both in worker threads but now also in threads that aren't tied to a script environment at all.

**Comment by gterzian** [comment]
Another update on the ongoing experiment. 

One problem we ran into was in the communication between layout and the font-cache, which in some cases goes via a blocking ipc call, implemented through creating a channel, sending the sender to the font-cache, and blocking on the receiver. 

This was a new usecase since the previous fixes all related to the use of `ROUTER` in script, in effect non-blocking ipc-calls. 

So what I did is add a `add_blocking_call` API to the  shared-router, which returns a `(IpcHandle<T>, Receiver<Vec<u8>>)`, where the receiver is a crossbeam receiver. 

The client can then "block" on the receiver, as opposed to passing a callback to the router. 

it looks like:

![Screen Shot 2019-08-08 at 1 42 51 PM](https://user-images.githubusercontent.com/2792687/62677907-73ea6780-b9e2-11e9-977f-0a98a600285d.png)

**Comment by gterzian** [comment]
Current problems to note:

1. Shared memory is not supported, hence it cannot be applied to the image-cache. 

The reason is that what is now sent is a `Callback(IpcCallbackId, Vec<u8>)`, where the `Vec<u8>` is the underlying message(for example a `FetchResponseMsg`), which is serialized first. 

So this first of all means you need two serialization per message, since you first serialize the "inner" message manually, and then send the "outer" one which will then be serialized again as a whole by the channel. 

If the "inner" message contains shared memory, then serializing it "by hand" outside of the call to `send` on the channel, breaks shared-memory. The reason seems to be that the channel internally manipulates the shared-mem stuff first, then serializes the message which calls into `serialize` of the shared memory. So basically you cannot serialize shared-memory outside of the actual send operation(I tried hacking together a "serializer" hook to `send`, it didn't work, probably because I didn't do anything on the receiving side).

2. Related to the above, is obviously there are a lot of calls to `bincode::deserialize` throwns around the "client" code, although the handle has some phantomdata so at least it's tied to a particular type of "inner" message. It would be nice if the whole thing could be typed, the main problem is actually circular dependencies, more precisely the font-cache message type is not in `gfx_traits`, and it's hard to move it to there(and since `gfx` uses the router, the router cannot be aware of types in `gfx`.).

**Comment by gterzian** [comment]
What's nice to note is that this gives us a single router per process, so you can imagine that all ipc in a process would go over a single ipc-channel. The expection is channels used "at the top of the loop", for example a script-thread receiving messages from the constellation. But that doesn't matter, since those channels are created once and used for all messages in that relationship. 

What goes away is all the stuff like "create an ipc-channel to do one fetch, or handle one websocket, load one stylsheet" and so on. 

It's basically a variation on the `ROUTER` in ipc-channel, that uses a single channel for all routes. and which provides one `IpcHandle` per route for sending messages to the corresponding route/callback, as opposed to requiring sending an actual `IpcSender` around.

Also, the fix for the image-cache problem might just be copy/pasting some code and having a dedicated router for the image-cache, which would use actual `ImageResponse` as opposed to `Vec<u8>`, allowing for shared-memory(since there would be no need to manually serialize the `ImageResponse` to include it in a `Callback(IpcCallbackId, Vec<u8>)`).

**Comment by gterzian** [comment]
So, I think it's time to review the experiment so far. 

On the problem side:

1. We create a lot of ipc-channels to implement cross-process "callbacks". On linux this can crash due to "too many open files", not sure what it does on other platform but I assume it will lead to problems across the board. 
2. My perspective is that the problem is not so much sending `IpcSender` around, which results in an extra fd on each send, see https://github.com/servo/servo/pull/23909. The problem rather is the creation of channels to implement callbacks, which requires a dedicated receiver since the callback is specific to the code creating the channel. 
3. Note that those "callbacks" often need to retain some state, and are used to handle more than one ipc-message. Example is a fetch, which sets-up a route using a newly created channel, and then handle various fetch response messages corresponding to the lifecycle progress of a fetch. The problem there is really that each fetch requires it's own channel and route.

Prototyped solution:

1. Use a `SharedRouter`, which internally sets-up a single route using a single channel. Clients can then "register a callback", in the form of a closure, and get a "handle" in return which they can send across process to the process from which they expect a reply. The "handle" itself doesn't have to contain an IpcSender, so it doesn't create an FD each time it is send over ipc. Instead, the shared-router needs to be setup specifically with each component that will receive such handles, effectively just by communicating an ipc-sender upon setup(as opposed to inside each handle). 
2. In terms of API, it works pretty well, since the current use of `ROUTER` already essentially consists of passing a closure to the route. So "registering a callback" and getting a "handle" in return is not that different from creating an ipc-channel and registering a route, passing the sender to the other process. 
3. I was able to refactor almost all use of `ROUTER` across Servo, including across components other than script, such as the constellation(In it;s own use of `NetworkListener`), and the font-cache(in it's use of fetch), as well as layout(in it's use of the font-cache). 
4. I was also pretty easy to have a component like the font-cache both use a router for it's own uses(fetch), as well as provide a "routing service" to another component, in this case layout. 
5. It does seem to prevent specific "too many open files" crashes, although each fix so far as lead to discovering new issues. 

Problems with the prototype:

1. It relies on double (de-)serialization of each message, since each message contains an "inner" message that is separately serialized and added to the "outer" callback message as a `Vec<u8>`. 
2. The double serialization, or rather serialization outside of the underlying channel, breaks any shared-memory contained in the message. The current implementation of shared-memory in ipc-channel requires the memory to be serialized inside the call to `send`. Hence, I wasn't able to apply the prototype to components using shared-memory, like the image cache. 
3. It generally adds a lot of calls to `bincode` across the codebase, since the callbacks are typed to a `Vec<u8>` and then manually deserialize it into the expected message. The main reason for this is that it's hard to type the whole thing without running into cyclic dependencies. It could be solved, or is already solveable if you only use the router from  a subset of components, like script, for a subset of messages, like fetch. Using it across Servo means you run into cyclic dependencies between `msg` and the crate using the router. 
4. There is generally quite a lot of glue code everywhere, since the router needs to be manually setup with components like fetch. 

My conclusion:

1. At the API level it works. Our current use of `ROUTER` can be replaced by some sort of callback based API. 
2. It's going to be probably too much work to make the prototype something that is production ready, like supporting shared mem and other high-performance features This would also require changes to ipc-channel. 
3. The glue code is a bit much. 

So, I think the best use of resources, would be writing Rust bindings to https://chromium.googlesource.com/chromium/src/+/master/mojo/public/cpp/bindings/README.md

It would allow sutff like:

```c++
void OnGetTail(const std::string& message) {
  LOG(ERROR) << "Tail was: " << message;
}

logger->GetTail(base::BindOnce(&OnGetTail));
```

It would also support much more advanced communication patterns, see for examples: https://chromium.googlesource.com/chromium/src/+/master/mojo/public/cpp/bindings/README.md#sending-interfaces-over-interfaces

We could still use a ipc-channel for "long-running" connections, like a single channel between the constellation and a script-process, since those channels are only created once in a while, as opposed to once for each operation. 

Over time, we could probably replace those as well. It could bring additional benefits to the modularity of the code, since we'd have to define clear interfaces that abstract the underlying ipc communication. 

Finally, it would give us something that works on Windows, and that is I assume well maintained and performant. The added value for Rust would really be the bindings, as opposed to re-writing the underlying plumbing(which consists mostly of unsafe calls to `libc`, at least on linux, hence a well-maintained C libary can provide the same value as a Rust implementation).

cc @jdm @asajeffrey

**Comment by jdm** [comment]
I'm not sure I follow why transitioning to using Rust bindings around Mojo is the logical conclusion. What is the problem being solved, and how does switching to Mojo accomplish that? At what level of the platform do you propose that switch should take place?

**Comment by asajeffrey** [comment]
Thanks for doing the experiments.

As well as the chromium IPC, one thing that we should investigate is whether we can make more use of shared memory directly, to save the serialization/deserialization costs of IPC. There is already the `shared_memory` crate that has a trait for data which can be safely put in shared memory: https://docs.rs/shared_memory/0.8.2/shared_memory/trait.SharedMemCast.html This is missing a `derive` implementation, and a way to share variable-length data like vectors.

One advantage of this over mojo would be that more of the code would be in Rust, which gives us memory safety guarantees (assuming we get the low-level bindings right).

**Comment by gterzian** [comment]
> What is the problem being solved, and how does switching to Mojo accomplish that?

@jdm 

The problem as I see it is that we want to use IPC for cheap callbacks, we're currently doing it by using the `ROUTER` and newly created ipc-channels, and it's not cheap. 

What we're missing is a cheap "handle" that can be passed around processes to send a message back to a callback in another process. Sending an `IpcSender` around means an FD on each receive, plus we need to actually create a channel since a callback needs a dedicated receiver to match the sender we're passing around. 

Mojo could solve that problem, since their API appears to fit what we need to do, and I assume the implementation doesn't strain resources like our use of the router/channels do. I need to check if that is correct actually. 

Mojo seems to internally use "handles" as internal endpoints which I assume are cheaper than file descriptors, see https://chromium.googlesource.com/chromium/src/+/master/mojo/public/c/system/README.md#handles

> At what level of the platform do you propose that switch should take place?

If you mean at what level of Servo, I think it would be most useful in places were we currently use the `ROUTER` with a newly created ipc-channel repeatedly, essentially the places where I started using the prototype shared-router.

**Comment by gterzian** [comment]
> As well as the chromium IPC, one thing that we should investigate is whether we can make more use of shared memory directly, to save the serialization/deserialization costs of IPC. 

@asajeffrey 

While I think the main bottleneck we're facing is not the serialization, it's rather the cost of creating and passing channels around(which map directly to system resources like file descriptors), having the capacity to further optimize things with shared-memory would certainly be useful. 

I have to say the support for that in Mojo seems outstanding, they even support stuff like https://chromium.googlesource.com/chromium/src/+/master/mojo/public/c/system/README.md#two_phase-i_o which looks like it could be really useful for stuff like webgl(this is somewhat naive since I don't actually know much about our use of shared buffers in webgl) 
cc @jdm 

> One advantage of this over mojo would be that more of the code would be in Rust, which gives us memory safety guarantees (assuming we get the low-level bindings right).

I somewhat buy the memory safety part less in this case, since the underlying implementation would be almost entirely unsafe(the trait you linked to is unsafe, for example). 

I would be much more comfortable with safe rust bindings to a battle-tested C library like Mojo. You're basically getting free support from Google in that case, and a lot of usage that I assume keeps the underlying implementation relatively bug free over time. 

Also, let's not forget about windows, and other platforms, support. Altough it looks like it's forthcoming for ipc-channel as well https://github.com/servo/ipc-channel/pull/233

**Comment by gterzian** [comment]
@jdm @asajeffrey Also, please do take a look at https://github.com/servo/servo/pull/23909 and let me know what you think. 

It essentially works, and there is some value in there if only because the ipc-stuff is isolated in one place. The "cheap" handle part works. 

The problem is really avoiding the double serialization, and supporting shared-memory containing messages.

**Comment by gterzian** [comment]
As a last argument, note that we're regularly adding code that uses ipc-channel as cheap callbacks

Example: https://github.com/servo/media/pull/289/files#diff-9b2c72cf27eb297db09cd71e2fd6d8fdR52 (cc @ferjm )

or my own ongoing work with message-ports(where I recently tried to at least only use one route per global, due to the latest experience). 

It seems like there is a strong need for doing things with ipc-channel that imply it's really cheap to create and pass around channels. And that seems to be what the API of Mojo is all about, hence I think it's an interesting prospect.

**Comment by gterzian** [comment]
To answer my own question about whether the Mojo stuff is expensive: 

> Are message pipes expensive?

> No. As an implementation detail, creating a message pipe is essentially generating two random numbers and stuffing them into a hash table, along with a few tiny heap allocations.

https://chromium.googlesource.com/chromium/src/+/master/mojo/README.md#are-message-pipes-expensive

**Comment by asajeffrey** [comment]
I did some messing around, and I think we can get memory-safe access to shared memory. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d8a88d07419ef92008075b4b340d9d9c is a first draft, which passes Miri.

On IRC, @jdm points out that there may be performance issues with shared memory vs ipc, unfortunately the only way to find out is to run some experiments.

I agree @gterzian that there's a tricky trade-off here between the battle-tested and Google-backed Mojo vs something new written in Rust. It's not obvious what the right thing to do is, I'm reminded of @avadacatavra's investigations of openSSL vs using a Rust TLS implementation. One difference is that we've already got a Rust IPC implementation, so in this case we'd be replacing an existing Rust library with a C++ one. @pcwalton may have opinions about that.

**Comment by gterzian** [comment]
This is an old doc from Chromium which discusses options to get close to zero-copy when transferring(a shared array buffer) over IPC, pre-Mojo. 

Might be relevant for strategies to do stuff with shared-memory with ipc-channel. 

https://docs.google.com/document/d/17o_cjtc3Gk5cDZhoUc37EyhhvBb4wMlCROJvoC2Aokg/edit#heading=h.pojdlcrtjok7

Summary: Currently MessagePort does not support sending a transferable ArrayBuffer by move semantics yet because Chromium's multi-process architecture makes it difficult to support zero-copy transfer (move semantics) over renderers. This doc proposes 4 options for transfering an ArrayBuffer over processes by (pseudo) move semantics.

It's more about avoiding copying, not about memory safety. I will look into the linked-to playground.

Question: if we `bincode::serizalize` something into a `Vec<u8>`, would it then save us anything to put it into `IpcSharedMemory` and send that on the channel, and read from it and deserialize it on the other side, versus sending the vec on the channel? 

And to avoid all copies would the object itself need to contain shared-mem, and then we'd just have to serialize some identifier and send a clone of the shared-mem directly on the channel? Avoiding any serialization of data?

**Comment by asajeffrey** [comment]
I did some rough-and-ready experimentation with using shared memory and condition variables rather than sockets as the implementation of ipc channels. The code is at https://github.com/asajeffrey/ipc-experiments/blob/master/examples/experiment.rs

It's not memory-safe, doesn't support sending channels over channels, is currently lossy and has a fixed-size buffer. This is all unrealistic, but good enough to do some preliminary timing. Sending a million messages took 2.16s compared to 4.89s using ipc-channels. This is quite promising!

**Comment by asajeffrey** [comment]
@gterzian yes, in order to get zero copy messaging, we'd have to allocate (e.g.) vectors directly in shared memory, and keep them in shared memory during their lifetime. Rather annoyingly this means that we'd have `SharedVec<T>` rather than `Vec<T>` (and ditto hash maps etc.) since pointers in shared memory have to be of the form offset-from-shared-page-base-address rather than just being a pointer.

**Comment by glyn** [comment]
Is the problem in the OP still happening, is it not encountered sufficiently frequently to force a resolution, has a workaround been developed, or what? Thanks for any update.

**Reply by jdm** [reply]
Yep, definitely still happening. https://github.com/servo/servo/issues/33672 is an example of it affecting real web content.
